xds/internal/server/rds_handler_test.go

```diff
@@ -351,51 +417,60 @@ func (s) TestSuccessCaseSecondUpdateMakesRouteFull(t *testing.T) {
-		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v", routeNameDeleted, route1)
-	}
-	rhuWant := map[string]xdsresource.RouteConfigUpdate{route1: {}, route2: {}}
-	select {
-	case rhu := <-ch:
-		if diff := cmp.Diff(rhu.updates, rhuWant); diff != "" {
-			t.Fatalf("got unexpected route update, diff (-got, +want): %v", diff)
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for the rds handler update to be written to the update buffer.")
+	wantUpdate := rdsHandlerUpdate{
+		updates: map[string]xdsresource.RouteConfigUpdate{
+			route1: defaultRouteConfigUpdate,
+			route2: defaultRouteConfigUpdate,
+		},
 	}
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{route1, route2})
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
+
+	// Close the rds handler and verify that the watch is canceled.
+	rh.close()
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{})
 }
 
 // TestErrorReceived tests the case where the rds handler receives a route name
 // to watch, then receives an update with an error. This error should be then
 // written to the update channel.
 func (s) TestErrorReceived(t *testing.T) {
-	rh, fakeClient, ch := setupTests()
+	mgmtServer, nodeID, resourceNamesCh, xdsC := setupForRDSHandlerTests(t)
 
+	// Create an rds handler and give it a single route to watch.
+	updateCh := make(chan rdsHandlerUpdate, 1)
+	rh := newRDSHandler(xdsC, updateCh)
 	rh.updateRouteNamesToWatch(map[string]bool{route1: true})
+
+	// Verify that the given route is requested.
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	gotRoute, err := fakeClient.WaitForWatchRouteConfig(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchRDS failed with error %v", err)
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{route1})
+
+	// Configure the management server with a single route config resource, that
+	// is expected to be NACKed.
+	routeResource := routeConfigResourceForName(route1)
+	routeResource.VirtualHosts[0].RetryPolicy = &v3routepb.RetryPolicy{NumRetries: &wrapperspb.UInt32Value{Value: 0}}
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Routes:         []*v3routepb.RouteConfiguration{routeResource},
+		SkipValidation: true,
 	}
-	if gotRoute != route1 {
-		t.Fatalf("xdsClient.WatchRDS called for route: %v, want %v", gotRoute, route1)
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
 	}
 
-	rdsErr := errors.New("some error")
-	fakeClient.InvokeWatchRouteConfigCallback(route1, xdsresource.RouteConfigUpdate{}, rdsErr)
+	const wantErr = "received route is invalid: retry_policy.num_retries = 0; must be >= 1"
 	select {
-	case rhu := <-ch:
-		if rhu.err.Error() != "some error" {
-			t.Fatalf("Did not receive the expected error, instead received: %v", rhu.err.Error())
+	case update := <-updateCh:
+		if !strings.Contains(update.err.Error(), wantErr) {
+			t.Fatalf("Update received with error %v, want error containing %q", update.err, wantErr)
```

Comment:
Why is the second formatting directive %q vs %v?

Reply:
Typo.